### DISCUSSION TITLE
Create slist of installed packages without using classic arrays

### DISCUSSION THREAD

**Discussion by olehermanse** [cfengine/core#5275] [Q&A]
I have the following policy:

```cfengine3
bundle agent find_installed_packages
{
  vars:
    "installed_packages_data"
      data => packagesmatching(
        ".*",
        ".*",
        ".*",
        "$(default:package_module_knowledge.platform_default)");
    "installed_packages_index"
      slist => getindices("installed_packages_data");
    "installed_packages_arr[$(installed_packages_index)]"
      string => "$(installed_packages_data[$(installed_packages_index)][name])";
    "installed_packages_names"
      slist => sort(getvalues(installed_packages_arr), "lex");
}
```

(Needs to be run with masterfiles / standard lib for package module variables to exist).

Which works, converting the results from `packagesmatching()` into a simple `slist` of package names of installed packages.

How can I achieve the same conversion without using a classic array? (Classic arrays have other quirks, like leaving behind variables when things change, i.e. when packages are uninstalled).

**Comment by nickanderson** [comment]
I was able to accomplish this without using classic associative arrays using `mapdata()`.

<https://docs.cfengine.com/docs/3.21/reference-functions-mapdata.html>

Here is a small example:

```cfengine3
bundle agent __main__
{
  vars:
      # Mock response from packagesmatching()
      "d" data => '[
   {
      "arch":"default",
      "method":"dpkg",
      "name":"zsh-common",
      "version":"5.0.7-5ubuntu1"
   },
   {
      "arch":"default",
      "method":"dpkg",
      "name":"emacs",
      "version":"1:27.1+1-3ubuntu5"
   }
]';

      # Use mapdata() to build a new data container where the keys are the pacakge name
      # We reference back to 'd' because this.k is the positional location in the array returned by packagesmatching()
      "_dl" data => mapdata( "none", '{ "$(d[$(this.k)][name])": "$(d[$(this.k)][name])" }', d);

      # Then use getvalues() to get the names of the keys
      # Yea, I wanted to use getindices first, but getindices() gives you the positional value
      "l" slist => getvalues( _dl );

  reports:
      "Packages: $(with)" with => join( ", ", l );
      "CFEngine version: $(sys.cf_version)";

}
```

    # cf-agent --no-lock --log-level info --show-evaluated-vars=default:main --file /tmp/example.cf
    R: Packages: { "zsh-common": "zsh-common" }, { "emacs": "emacs" }
    R: CFEngine version: 3.22.0
    Variable name                            Variable value                                               Meta tags                                Comment
    default:main._dl                         ["{ \"zsh-common\": \"zsh-common\" }","{ \"emacs\": \"emacs\" }"] source=promise
    default:main.d                           [{"arch":"default","method":"dpkg","name":"zsh-common","version":"5.0.7-5ubuntu1"},{"arch":"default","method":"dpkg","name":"emacs","version":"1:27.1+1-3ubuntu5"}] source=promise
    default:main.l                            {"{ \"zsh-common\": \"zsh-common\" }","{ \"emacs\": \"emacs\" }"} source=promise

**Reply by nickanderson** [reply]
OOps, I guess not. There your not getting a list. Boo Dyslexia

**Reply by nickanderson** [reply]
Ah, you can get to it with `mapdata()`.

```cfengine3
bundle agent __main__
{
  vars:
      # Mock response from packagesmatching()
      "d" data => '[
   {
      "arch":"default",
      "method":"dpkg",
      "name":"zsh-common",
      "version":"5.0.7-5ubuntu1"
   },
   {
      "arch":"default",
      "method":"dpkg",
      "name":"emacs",
      "version":"1:27.1+1-3ubuntu5"
   }
]';

      # Use mapdata() to build a new data container where the keys are the pacakge name
      # We reference back to 'd' because this.k is the positional location in the array returned by packagesmatching()
      "_dl" data => mapdata( "none", '$(d[$(this.k)][name])', d);

      # Then use getvalues() to get the names of the keys
      # Yea, I wanted to use getindices first, but getindices() gives you the positional value
      "l" slist => sort( getvalues( _dl ), lex);

  reports:
      "Packages: $(with)" with => join( ", ", l );
      "CFEngine version: $(sys.cf_version)";

}
```

    # cf-agent --no-lock --log-level info --show-evaluated-vars=default:main --file /tmp/example.cf
    R: Packages: emacs, zsh-common
    R: CFEngine version: 3.22.0
    Variable name                            Variable value                                               Meta tags                                Comment
    default:main._dl                         ["zsh-common","emacs"]                                       source=promise
    default:main.d                           [{"arch":"default","method":"dpkg","name":"zsh-common","version":"5.0.7-5ubuntu1"},{"arch":"default","method":"dpkg","name":"emacs","version":"1:27.1+1-3ubuntu5"}] source=promise
    default:main.l                            {"emacs","zsh-common"}                                      source=promise

**Comment by nickanderson** [comment]
For real this time, I was able to accomplish this without using classic associative arrays using `maparray()`.

<https://docs.cfengine.com/docs/3.21/reference-functions-maparray.html>

Here is a small example:

```cfengine3
bundle agent __main__
{
  vars:
      # Mock response from packagesmatching()
      "d" data => '[
   {
      "arch":"default",
      "method":"dpkg",
      "name":"zsh-common",
      "version":"5.0.7-5ubuntu1"
   },
   {
      "arch":"default",
      "method":"dpkg",
      "name":"emacs",
      "version":"1:27.1+1-3ubuntu5"
   }
]';

      # Use mapdata() to build a new data container where the keys are the pacakge name
      # We reference back to 'd' because this.k is the positional location in the array returned by packagesmatching()
      # "_dl" data => mapdata( "none", '[ "$(d[$(this.k)][name])" ]', d);
      "_dl" slist => maparray( '$(d[$(this.k)][name])', d);

      # Then use getvalues() to get the names of the keys
      # Yea, I wanted to use getindices first, but getindices() gives you the positional value
      "l" slist => sort( getvalues( _dl ), lex);

  reports:
      "Packages: $(with)" with => join( ", ", l );
      "CFEngine version: $(sys.cf_version)";

}
```

    # cf-agent --no-lock --log-level info --show-evaluated-vars=default:main --file /tmp/example.cf
    R: Packages: emacs, zsh-common
    R: CFEngine version: 3.22.0
    Variable name                            Variable value                                               Meta tags                                Comment
    default:main._dl                          {"zsh-common","emacs"}                                      source=promise
    default:main.d                           [{"arch":"default","method":"dpkg","name":"zsh-common","version":"5.0.7-5ubuntu1"},{"arch":"default","method":"dpkg","name":"emacs","version":"1:27.1+1-3ubuntu5"}] source=promise
    default:main.l                            {"emacs","zsh-common"}                                      source=promise
